plugins/algoltek-usbcr/fu-algoltek-usbcr-device.c

```diff
@@ -0,0 +1,671 @@
+{
+	guint8 buf[8] = {0};
+	g_autoptr(GByteArray) st = fu_struct_ag_usbcr_spi_cdb_new();
+
+	fu_struct_ag_usbcr_spi_cdb_set_cmd(st, FU_AG_USBCR_SCSIOP_VENDOR_EEPROM_WR);
+	fu_struct_ag_usbcr_spi_cdb_set_addr(st, 0xFFFF);
+	fu_struct_ag_usbcr_spi_cdb_set_bufsz(st, sizeof(buf) & 0xFF);
+	fu_struct_ag_usbcr_spi_cdb_set_tag(st, FU_AG_SPECIFY_EEPROM_TYPE_TAG);
+	fu_struct_ag_usbcr_spi_cdb_set_valid(st, FU_AG_SPIFLASH_VALID);
+	fu_struct_ag_usbcr_spi_cdb_set_spisig1(st, FU_AG_SPECIFY_SPI_CMD_SIG_1);
+	fu_struct_ag_usbcr_spi_cdb_set_spisig2(st, FU_AG_SPECIFY_SPI_CMD_SIG_2);
+	fu_struct_ag_usbcr_spi_cdb_set_spicmd(st, cmd);
+
+	return fu_block_device_sg_io_cmd_write(FU_BLOCK_DEVICE(self),
+					       st->data,
+					       st->len,
+					       buf,
+					       sizeof(buf),
+					       error);
+}
+
+static gboolean
+fu_algoltek_usbcr_device_do_write_spi(FuAlgoltekUsbcrDevice *self,
+				      guint16 addr,
+				      const guint8 *buf,
+				      guint8 bufsz,
+				      guint8 access_sz,
+				      GError **error)
+{
+	g_autoptr(GByteArray) st = fu_struct_ag_usbcr_spi_cdb_new();
+
+	if (!fu_algoltek_usbcr_device_send_spi_cmd(self, FU_AG_USBCR_WREN, error))
+		return FALSE;
+
+	fu_struct_ag_usbcr_spi_cdb_set_cmd(st, FU_AG_USBCR_SCSIOP_VENDOR_EEPROM_WR);
+	fu_struct_ag_usbcr_spi_cdb_set_addr(st, addr);
+	fu_struct_ag_usbcr_spi_cdb_set_bufsz(st, access_sz);
+	fu_struct_ag_usbcr_spi_cdb_set_tag(st, FU_AG_SPECIFY_EEPROM_TYPE_TAG);
+	fu_struct_ag_usbcr_spi_cdb_set_valid(st, FU_AG_SPIFLASH_VALID);
+
+	return fu_block_device_sg_io_cmd_write(FU_BLOCK_DEVICE(self),
+					       st->data,
+					       st->len,
+					       buf,
+					       bufsz,
+					       error);
+}
+
+static gboolean
+fu_algoltek_usbcr_device_do_read_spi(FuAlgoltekUsbcrDevice *self,
+				     guint16 addr,
+				     guint8 *buf,
+				     guint8 bufsz,
+				     GError **error)
+{
+	g_autoptr(GByteArray) st = fu_struct_ag_usbcr_spi_cdb_new();
+
+	fu_struct_ag_usbcr_spi_cdb_set_cmd(st, FU_AG_USBCR_SCSIOP_VENDOR_EEPROM_RD);
+	fu_struct_ag_usbcr_spi_cdb_set_addr(st, addr);
+	fu_struct_ag_usbcr_spi_cdb_set_bufsz(st, bufsz);
+	fu_struct_ag_usbcr_spi_cdb_set_tag(st, FU_AG_SPECIFY_EEPROM_TYPE_TAG);
+	fu_struct_ag_usbcr_spi_cdb_set_valid(st, FU_AG_SPIFLASH_VALID);
+
+	return fu_block_device_sg_io_cmd_read(FU_BLOCK_DEVICE(self),
+					      st->data,
+					      st->len,
+					      buf,
+					      bufsz,
+					      error);
+}
+
+static gboolean
+fu_algoltek_usbcr_device_command_wren(FuAlgoltekUsbcrDevice *self, GError **error)
+{
+	guint8 buf[1] = {0};
+	FuAgUsbcrRegSetup regs[] = {
+	    {0xC8, 0x04},
+	    {0xCA, 0x01},
+	    {0x400, FU_AG_USBCR_WREN},
+	    {0xC9, 0x01},
+	    {0xC8, 0x05},
+	};
+
+	for (guint i = 0; i < G_N_ELEMENTS(regs); i++) {
+		if (!fu_algoltek_usbcr_device_write_reg(self,
+							regs[i].reg,
+							regs[i].val,
+							FU_AG_USBCR_RD_WR_XDATA,
+							error))
+			return FALSE;
+	}
+	do {
```

Comment:
i think this needs to use fu_device_retry too

Reply:
Updated, use fu_device_retry_full() instead of do-while.